amethyst_renderer/src/pipe/effect/mod.rs

```diff
@@ -0,0 +1,201 @@
+//! Helper builder for pipeline state objects.
+
+#![allow(missing_docs)]
+
+use self::pso::{Data, Init, Meta};
+
+use error::{Error, Result};
+use fnv::FnvHashMap as HashMap;
+use gfx::{Primitive, ShaderSet};
+use gfx::preset::depth::{LESS_EQUAL_TEST, LESS_EQUAL_WRITE};
+use gfx::shade::{ProgramError, ToUniform};
+use gfx::state::{Depth, Rasterizer};
+use gfx::texture::{FilterMethod, SamplerInfo, WrapMode};
+use pipe::{Target, Targets};
+use types::{Factory, PipelineState, Resources, Sampler};
+
+mod pso;
+
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub enum BlendMode {
+    Add,
+    Alpha,
+    Invert,
+    Multiply,
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub enum DepthMode {
+    LessEqualTest,
+    LessEqualWrite,
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+enum ProgramSource {
+    Simple(&'static [u8], &'static [u8]),
+    Geometry(&'static [u8], &'static [u8], &'static [u8]),
+    Tessellated(&'static [u8], &'static [u8], &'static [u8], &'static [u8]),
+}
+
+impl ProgramSource {
+    pub fn compile(&self, fac: &mut Factory) -> Result<ShaderSet<Resources>> {
+        use gfx::Factory;
+        use gfx::traits::FactoryExt;
+
+        match *self {
+            ProgramSource::Simple(ref vs, ref ps) => {
+                fac.create_shader_set(vs, ps)
+                    .map_err(|e| Error::ProgramCreation(e))
+            }
+            ProgramSource::Geometry(ref vs, ref gs, ref ps) => {
+                let v = fac.create_shader_vertex(vs)
+                    .map_err(|e| ProgramError::Vertex(e))?;
+                let g = fac.create_shader_geometry(gs)
+                    .expect("Geometry shader creation failed");
+                let p = fac.create_shader_pixel(ps)
+                    .map_err(|e| ProgramError::Pixel(e))?;
+                Ok(ShaderSet::Geometry(v, g, p))
+            }
+            ProgramSource::Tessellated(ref vs, ref hs, ref ds, ref ps) => {
+                fac.create_shader_set_tessellation(vs, hs, ds, ps)
+                    .map_err(|e| Error::ProgramCreation(e))
+            }
+        }
+    }
+}
+
+#[derive(Derivative)]
+#[derivative(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct Effect {
+    pso: PipelineState<Meta>,
+    #[derivative(Hash = "ignore")]
+    pso_data: Data,
+    #[derivative(Hash = "ignore")]
+    samplers: HashMap<String, Sampler>,
+}
+
+impl Effect {
+    pub fn new_simple_prog<S>(vs: S, ps: S) -> EffectBuilder
```

Comment:
Same as for `Mesh`. Please move these functions to the builder type, I think that's clearer.